presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -399,27 +418,136 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+
+        public void addBlocksToHash(GroupByIdBlock groupIds, Block[] blocks)
+        {
+            Work<Block> work = hash.markDistinctRows(new Page(blocks));
+            checkState(work.process());
+            Block distinctMask = work.getResult();
+
+            blocks[0] = distinctMask;
+            accumulator.addInput(groupIds, new Page(blocks));
+        }
+    }
+
+    private static class SpillableDistinctingGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final DistinctingGroupedAccumulator distinctingGroupedAccumulator;
+        private final ObjectBigArray<RowBlockBuilder> intermediateState;
+        private final List<Type> spillingTypes;
+
+        public SpillableDistinctingGroupedAccumulator(GroupedAccumulator accumulator,
+                List<Type> inputTypes,
+                List<Type> sourceTypes,
+                List<Integer> inputChannels,
+                Optional<Integer> maskChannel,
+                Session session,
+                JoinCompiler joinCompiler)
+        {
+            this.distinctingGroupedAccumulator = new DistinctingGroupedAccumulator(accumulator, inputTypes, sourceTypes, inputChannels, maskChannel, session, joinCompiler);
+            this.spillingTypes = sourceTypes;
+            this.intermediateState = new ObjectBigArray<>();
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return distinctingGroupedAccumulator.getEstimatedSize() + intermediateState.sizeOf();
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return distinctingGroupedAccumulator.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            return new ArrayType(RowType.anonymous(spillingTypes));
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIds, Page page)
+        {
+            for (int pos = 0; pos < groupIds.getPositionCount(); pos++) {
+                long groupId = groupIds.getGroupId(pos);
+                intermediateState.ensureCapacity(groupId);
+                RowBlockBuilder rowBlockBuilder = intermediateState.get(groupId);
+                if (rowBlockBuilder == null) {
+                    rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIds.getGroupCount());
+                }
+
+                BlockBuilder output = rowBlockBuilder.beginBlockEntry();
+                for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                    spillingTypes.get(channel).appendTo(page.getBlock(channel), pos, output);
+                }
+                rowBlockBuilder.closeEntry();
+
+                intermediateState.set(groupId, rowBlockBuilder);
+            }
+            // TODO do this in prepareFinal() instead because this duplicates data in intermediateState and distincting hash table
+            distinctingGroupedAccumulator.addInput(groupIds, page);
```

Comment:
Should happen in prepareFinal() but unsure how to do without using twice as much memory, same issue as order by.